Calibration

adamb56789 opened the pull request:
The hub returns errors to all of the requests because it can't connect to the motor server, but the requests are still correct.

closes #112

nout-kleef commented:
Entering calibration mode works fine, but there are a couple of problems:

- ![Screenshot from 2021-03-16 16-12-35](https://user-images.githubusercontent.com/30589554/111342813-b706c500-8672-11eb-93d1-2ab40660996c.png)
  `move_down`, `move_up` and `stop_moving` requests time out
- ![Screenshot from 2021-03-16 16-17-28](https://user-images.githubusercontent.com/30589554/111343284-309eb300-8673-11eb-8910-244e8b9edae6.png)
  Similarly for `set_highest` and `set_lowest`
- ![Screenshot from 2021-03-16 16-12-44](https://user-images.githubusercontent.com/30589554/111342939-d6055700-8672-11eb-91ca-2b6e8f712773.png)
  Exiting the calibration page in the app doesn't trigger `/motor/calibrate/stop`

Any idea what's wrong? My `helio-hub` clone is up to date, so I don't think that's the issue.

nout-kleef commented:
A few minutes after `set_highest` and `set_lowest`, the following happens:
![Screenshot from 2021-03-16 16-19-26](https://user-images.githubusercontent.com/30589554/111343647-87a48800-8673-11eb-9b2e-c233cabbc274.png)
Is this a Hub-side issue?

adamb56789 commented:
I fixed stop not being triggered, as well as start being triggered incorrectly + some other stuff. For the other issues, I have no idea because that isn't happening on my machine. I get error 500 back from the server instead of timeout.

Here are some logs. The procedure was
1. Restart server and app
1. Click edit blinds
2. Open a blind
3. Click calibration
4. Hold down move up for a short time
5. Press set high point
6. Hold down move down for a short time
7. Press set low point
8. Press back

### App

```
03/16 19:31:37: Launching 'app' on Pixel_3a_API_30_x86.
Install successfully finished in 416 ms.
$ adb shell am start -n "com.helio.app/com.helio.app.MainActivity" -a android.intent.action.MAIN -c android.intent.category.LAUNCHER
Connected to process 27975 on device 'Pixel_3a_API_30_x86 [emulator-5554]'.
Capturing and displaying logcat messages from application. This behavior can be disabled in the "Logcat output" section of the "Debugger" settings page.
I/com.helio.app: Not late-enabling -Xcheck:jni (already on)
I/com.helio.app: Unquickening 12 vdex files!
W/com.helio.app: Unexpected CPU variant for X86 using defaults: x86
D/NetworkSecurityConfig: No Network Security Config specified, using platform default
D/NetworkSecurityConfig: No Network Security Config specified, using platform default
D/libEGL: loaded /vendor/lib/egl/libEGL_emulation.so
D/libEGL: loaded /vendor/lib/egl/libGLESv1_CM_emulation.so
D/libEGL: loaded /vendor/lib/egl/libGLESv2_emulation.so
W/com.helio.app: Accessing hidden method Landroid/view/View;->computeFitSystemWindows(Landroid/graphics/Rect;Landroid/graphics/Rect;)Z (greylist, reflection, allowed)
W/com.helio.app: Accessing hidden method Landroid/view/ViewGroup;->makeOptionalFitsSystemWindows()V (greylist, reflection, allowed)
I/okhttp.OkHttpClient: --> GET http://10.0.2.2:4310/motor/get_all
    --> END GET
I/okhttp.OkHttpClient: <-- 200 OK http://10.0.2.2:4310/motor/get_all (38ms)
    Content-Type: application/json
    Content-Length: 519
I/okhttp.OkHttpClient: Server: Werkzeug/1.0.1 Python/3.9.1
    Date: Tue, 16 Mar 2021 19:31:38 GMT
I/okhttp.OkHttpClient: [
I/okhttp.OkHttpClient:   {
        "active": true,
        "battery": 0,
        "id": 1,
        "ip": "0.0.0.0",
        "length": 0,
        "level": 20,
        "name": "\u00e3\u0192\u2020\u00e3\u201a\u00b9\u00e3\u0192\u02c6",
        "style": ""
      },
I/okhttp.OkHttpClient:   {
        "active": true,
        "battery": 0,
        "id": 3,
        "ip": "0.0.0.0",
        "length": 0,
        "level": 0,
I/okhttp.OkHttpClient:     "name": "\u30c6\u30b9\u30c8",
        "style": ""
      },
      {
        "active": true,
        "battery": 0,
        "id": 4,
        "ip": "0.0.0.0",
        "length": 0,
        "level": 45,
        "name": "test",
        "style": ""
      }
    ]
I/okhttp.OkHttpClient: <-- END HTTP (519-byte body)
D/HostConnection: HostConnection::get() New Host Connection established 0xf4e7bd50, tid 28015
D/HostConnection: HostComposition ext ANDROID_EMU_CHECKSUM_HELPER_v1 ANDROID_EMU_native_sync_v2 ANDROID_EMU_native_sync_v3 ANDROID_EMU_native_sync_v4 ANDROID_EMU_dma_v1 ANDROID_EMU_direct_mem ANDROID_EMU_host_composition_v1 ANDROID_EMU_host_composition_v2 ANDROID_EMU_vulkan ANDROID_EMU_deferred_vulkan_commands ANDROID_EMU_vulkan_null_optional_strings ANDROID_EMU_vulkan_create_resources_with_requirements ANDROID_EMU_YUV_Cache ANDROID_EMU_async_unmap_buffer ANDROID_EMU_vulkan_ignored_handles ANDROID_EMU_vulkan_free_memory_sync ANDROID_EMU_vulkan_shader_float16_int8 ANDROID_EMU_vulkan_async_queue_submit GL_OES_vertex_array_object GL_KHR_texture_compression_astc_ldr ANDROID_EMU_host_side_tracing ANDROID_EMU_async_frame_commands ANDROID_EMU_gles_max_version_2 
W/OpenGLRenderer: Failed to choose config with EGL_SWAP_BEHAVIOR_PRESERVED, retrying without...
D/EGL_emulation: eglCreateContext: 0xf4e7d720: maj 2 min 0 rcv 2
D/EGL_emulation: eglMakeCurrent: 0xf4e7d720: ver 2 0 (tinfo 0xf51bd1d0) (first time)
I/Gralloc4: mapper 4.x is not supported
D/HostConnection: createUnique: call
D/HostConnection: HostConnection::get() New Host Connection established 0xf4e7e280, tid 28015
D/goldfish-address-space: allocate: Ask for block of size 0x100
    allocate: ioctl allocate returned offset 0x3fb3bb000 size 0x2000
D/HostConnection: HostComposition ext ANDROID_EMU_CHECKSUM_HELPER_v1 ANDROID_EMU_native_sync_v2 ANDROID_EMU_native_sync_v3 ANDROID_EMU_native_sync_v4 ANDROID_EMU_dma_v1 ANDROID_EMU_direct_mem ANDROID_EMU_host_composition_v1 ANDROID_EMU_host_composition_v2 ANDROID_EMU_vulkan ANDROID_EMU_deferred_vulkan_commands ANDROID_EMU_vulkan_null_optional_strings ANDROID_EMU_vulkan_create_resources_with_requirements ANDROID_EMU_YUV_Cache ANDROID_EMU_async_unmap_buffer ANDROID_EMU_vulkan_ignored_handles ANDROID_EMU_vulkan_free_memory_sync ANDROID_EMU_vulkan_shader_float16_int8 ANDROID_EMU_vulkan_async_queue_submit GL_OES_vertex_array_object GL_KHR_texture_compression_astc_ldr ANDROID_EMU_host_side_tracing ANDROID_EMU_async_frame_commands ANDROID_EMU_gles_max_version_2 
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@6b041bb succeeded: [Motor{id='1', name='ãƒ†ã‚¹ãƒˆ', ip='0.0.0.0', active=true, level=20, battery=0, length=0, style='', icon=null}, Motor{id='3', name='テスト', ip='0.0.0.0', active=true, level=0, battery=0, length=0, style='', icon=null}, Motor{id='4', name='test', ip='0.0.0.0', active=true, level=45, battery=0, length=0, style='', icon=null}]
    Updated local state for [Motor{id='1', name='ãƒ†ã‚¹ãƒˆ', ip='0.0.0.0', active=true, level=20, battery=0, length=0, style='', icon=null}, Motor{id='3', name='テスト', ip='0.0.0.0', active=true, level=0, battery=0, length=0, style='', icon=null}, Motor{id='4', name='test', ip='0.0.0.0', active=true, level=45, battery=0, length=0, style='', icon=null}]
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/start/4
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- 200 OK http://10.0.2.2:4310/motor/calibrate/start/4 (7ms)
    Content-Type: application/json
    Content-Length: 22
    Server: Werkzeug/1.0.1 Python/3.9.1
    Date: Tue, 16 Mar 2021 19:32:04 GMT
    "Calibration started"
I/okhttp.OkHttpClient: <-- END HTTP (22-byte body)
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/update/4
D/CompatibilityChangeReporter: Compat change id reported: 147798919; UID 10153; state: ENABLED
I/okhttp.OkHttpClient: Content-Type: application/json; charset=UTF-8
    Content-Length: 89
    {"active":true,"battery":0,"ip":"0.0.0.0","length":0,"level":45,"name":"test","style":""}
    --> END PATCH (89-byte body)
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@9d9078f succeeded: Response{protocol=http/1.0, code=200, message=OK, url=http://10.0.2.2:4310/motor/calibrate/start/4}
I/okhttp.OkHttpClient: <-- 200 OK http://10.0.2.2:4310/motor/update/4 (17ms)
I/okhttp.OkHttpClient: Content-Type: application/json
    Content-Length: 130
    Server: Werkzeug/1.0.1 Python/3.9.1
    Date: Tue, 16 Mar 2021 19:32:04 GMT
    {
      "active": true,
      "battery": 0,
      "id": 4,
      "ip": "0.0.0.0",
      "length": 0,
I/okhttp.OkHttpClient:   "level": 45,
      "name": "test",
      "style": ""
    }
    <-- END HTTP (130-byte body)
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@6709c08 succeeded: Motor{id='4', name='test', ip='0.0.0.0', active=true, level=45, battery=0, length=0, style='', icon=null}
    Updating local state for Motor{id='4', name='test', ip='0.0.0.0', active=true, level=45, battery=0, length=0, style='', icon=null}
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/move_up/4
    Content-Length: 0
I/okhttp.OkHttpClient: --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- 500 INTERNAL SERVER ERROR http://10.0.2.2:4310/motor/calibrate/move_up/4 (9ms)
    Content-Type: application/json
    Content-Length: 24
    Server: Werkzeug/1.0.1 Python/3.9.1
    Date: Tue, 16 Mar 2021 19:32:04 GMT
    "Internal server error"
I/okhttp.OkHttpClient: <-- END HTTP (24-byte body)
I/System.out: Communication error
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/stop_moving/4
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- 500 INTERNAL SERVER ERROR http://10.0.2.2:4310/motor/calibrate/stop_moving/4 (7ms)
    Content-Type: application/json
    Content-Length: 24
    Server: Werkzeug/1.0.1 Python/3.9.1
    Date: Tue, 16 Mar 2021 19:32:05 GMT
    "Internal server error"
    <-- END HTTP (24-byte body)
I/System.out: Communication error
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/set_highest/4
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- 500 INTERNAL SERVER ERROR http://10.0.2.2:4310/motor/calibrate/set_highest/4 (9ms)
    Content-Type: application/json
    Content-Length: 24
    Server: Werkzeug/1.0.1 Python/3.9.1
    Date: Tue, 16 Mar 2021 19:32:06 GMT
I/okhttp.OkHttpClient: "Internal server error"
    <-- END HTTP (24-byte body)
I/System.out: Communication error
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/move_down/4
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- 500 INTERNAL SERVER ERROR http://10.0.2.2:4310/motor/calibrate/move_down/4 (8ms)
    Content-Type: application/json
    Content-Length: 24
    Server: Werkzeug/1.0.1 Python/3.9.1
I/okhttp.OkHttpClient: Date: Tue, 16 Mar 2021 19:32:07 GMT
    "Internal server error"
    <-- END HTTP (24-byte body)
I/System.out: Communication error
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/stop_moving/4
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- 500 INTERNAL SERVER ERROR http://10.0.2.2:4310/motor/calibrate/stop_moving/4 (11ms)
    Content-Type: application/json
    Content-Length: 24
    Server: Werkzeug/1.0.1 Python/3.9.1
    Date: Tue, 16 Mar 2021 19:32:08 GMT
    "Internal server error"
    <-- END HTTP (24-byte body)
I/System.out: Communication error
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/set_lowest/4
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- 500 INTERNAL SERVER ERROR http://10.0.2.2:4310/motor/calibrate/set_lowest/4 (11ms)
    Content-Type: application/json
    Content-Length: 24
    Server: Werkzeug/1.0.1 Python/3.9.1
    Date: Tue, 16 Mar 2021 19:32:08 GMT
    "Internal server error"
    <-- END HTTP (24-byte body)
I/System.out: Communication error
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/stop_moving/4
I/okhttp.OkHttpClient: Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/stop/4
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- 200 OK http://10.0.2.2:4310/motor/calibrate/stop/4 (9ms)
    Content-Type: application/json
    Content-Length: 23
    Server: Werkzeug/1.0.1 Python/3.9.1
    Date: Tue, 16 Mar 2021 19:32:09 GMT
    "Calibration finished"
    <-- END HTTP (23-byte body)
I/okhttp.OkHttpClient: <-- 500 INTERNAL SERVER ERROR http://10.0.2.2:4310/motor/calibrate/stop_moving/4 (15ms)
    Content-Type: application/json
I/okhttp.OkHttpClient: Content-Length: 24
    Server: Werkzeug/1.0.1 Python/3.9.1
    Date: Tue, 16 Mar 2021 19:32:09 GMT
    "Internal server error"
    <-- END HTTP (24-byte body)
I/com.helio.app: JIT allocated 74KB for compiled code of void android.view.View.<init>(android.content.Context, android.util.AttributeSet, int, int)
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@8e23c6e succeeded: Response{protocol=http/1.0, code=200, message=OK, url=http://10.0.2.2:4310/motor/calibrate/stop/4}
    Communication error
```

### Hub

```
*********************  THREADS  **********************
MainThread
SchedulerThread
******************************************************
 * Serving Flask app "api" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: on
 * Restarting with stat
*********************  THREADS  **********************
MainThread
SchedulerThread
******************************************************
 * Debugger is active!
 * Debugger PIN: 251-416-516
 * Running on http://0.0.0.0:4310/ (Press CTRL+C to quit)
127.0.0.1 - - [16/Mar/2021 19:31:38] "←[37mGET /motor/get_all HTTP/1.1←[0m" 200 -
127.0.0.1 - - [16/Mar/2021 19:32:04] "←[37mPATCH /motor/calibrate/start/4 HTTP/1.1←[0m" 200 -
127.0.0.1 - - [16/Mar/2021 19:32:04] "←[37mPATCH /motor/update/4 HTTP/1.1←[0m" 200 -
HTTPConnectionPool(host='0.0.0.0', port=4310): Max retries exceeded with url: /calibration_move_up (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001E11DA88340>: Failed to establish a new connection: [WinError 10049] The requested address is not valid in its context'))
127.0.0.1 - - [16/Mar/2021 19:32:04] "←[35m←[1mPATCH /motor/calibrate/move_up/4 HTTP/1.1←[0m" 500 -
HTTPConnectionPool(host='0.0.0.0', port=4310): Max retries exceeded with url: /calibration_stop_moving (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001E11DA88940>: Failed to establish a new connection: [WinError 10049] The requested address is not valid in its context'))
127.0.0.1 - - [16/Mar/2021 19:32:05] "←[35m←[1mPATCH /motor/calibrate/stop_moving/4 HTTP/1.1←[0m" 500 -
HTTPConnectionPool(host='0.0.0.0', port=4310): Max retries exceeded with url: /calibration_set_highest (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001E11DAA28E0>: Failed to establish a new connection: [WinError 10049] The requested address is not valid in its context'))
127.0.0.1 - - [16/Mar/2021 19:32:06] "←[35m←[1mPATCH /motor/calibrate/set_highest/4 HTTP/1.1←[0m" 500 -
HTTPConnectionPool(host='0.0.0.0', port=4310): Max retries exceeded with url: /calibration_move_down (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001E11DAA2640>: Failed to establish a new connection: [WinError 10049] The requested address is not valid in its context'))
127.0.0.1 - - [16/Mar/2021 19:32:07] "←[35m←[1mPATCH /motor/calibrate/move_down/4 HTTP/1.1←[0m" 500 -
HTTPConnectionPool(host='0.0.0.0', port=4310): Max retries exceeded with url: /calibration_stop_moving (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001E11DA885B0>: Failed to establish a new connection: [WinError 10049] The requested address is not valid in its context'))
127.0.0.1 - - [16/Mar/2021 19:32:08] "←[35m←[1mPATCH /motor/calibrate/stop_moving/4 HTTP/1.1←[0m" 500 -
HTTPConnectionPool(host='0.0.0.0', port=4310): Max retries exceeded with url: /calibration_set_lowest (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001E11DAA77F0>: Failed to establish a new connection: [WinError 10049] The requested address is not valid in its context'))
127.0.0.1 - - [16/Mar/2021 19:32:08] "←[35m←[1mPATCH /motor/calibrate/set_lowest/4 HTTP/1.1←[0m" 500 -
127.0.0.1 - - [16/Mar/2021 19:32:09] "←[37mPATCH /motor/calibrate/stop/4 HTTP/1.1←[0m" 200 -
HTTPConnectionPool(host='0.0.0.0', port=4310): Max retries exceeded with url: /calibration_stop_moving (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001E11DA6AC70>: Failed to establish a new connection: [WinError 10049] The requested address is not valid in its context'))
127.0.0.1 - - [16/Mar/2021 19:32:09] "←[35m←[1mPATCH /motor/calibrate/stop_moving/4 HTTP/1.1←[0m" 500 -
```

nout-kleef commented:
`/motor/calibrate/stop` is now being triggered when I exit the calibration fragment, but the other problems persist (following the steps in your previous comment).
You mentioned during our group meeting that you added a `time.sleep(1)` to the `helio-hub` code. Can you try repeating the steps with a clean `helio-hub`?

For reference, these are the logs from my machine

```
03/16 21:42:43: Launching 'app' on Pixel XL API 25.
App restart successful without requiring a re-install.
$ adb shell am start -n "com.helio.app/com.helio.app.MainActivity" -a android.intent.action.MAIN -c android.intent.category.LAUNCHER
Connected to process 4567 on device 'Pixel_XL_API_25 [emulator-5554]'.
Capturing and displaying logcat messages from application. This behavior can be disabled in the "Logcat output" section of the "Debugger" settings page.
I/art: Not late-enabling -Xcheck:jni (already on)
W/art: Unexpected CPU variant for X86 using defaults: x86
W/System: ClassLoader referenced unknown path: /data/app/com.helio.app-1/lib/x86
W/art: Before Android 4.1, method android.graphics.PorterDuffColorFilter androidx.vectordrawable.graphics.drawable.VectorDrawableCompat.updateTintFilter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have incorrectly overridden the package-private method in android.graphics.drawable.Drawable
I/System.out: "2131820593" is not a valid theme id, loading default
I/art: Rejecting re-init on previously-failed class java.lang.Class<androidx.core.view.ViewCompat$2>: java.lang.NoClassDefFoundError: Failed resolution of: Landroid/view/View$OnUnhandledKeyEventListener;
        at void androidx.core.view.ViewCompat.setBackground(android.view.View, android.graphics.drawable.Drawable) (ViewCompat.java:2678)
        at void androidx.appcompat.widget.ActionBarContainer.<init>(android.content.Context, android.util.AttributeSet) (ActionBarContainer.java:63)
        at java.lang.Object java.lang.reflect.Constructor.newInstance0!(java.lang.Object[]) (Constructor.java:-2)
        at java.lang.Object java.lang.reflect.Constructor.newInstance(java.lang.Object[]) (Constructor.java:430)
        at android.view.View android.view.LayoutInflater.createView(java.lang.String, java.lang.String, android.util.AttributeSet) (LayoutInflater.java:645)
        at android.view.View android.view.LayoutInflater.createViewFromTag(android.view.View, java.lang.String, android.content.Context, android.util.AttributeSet, boolean) (LayoutInflater.java:787)
        at android.view.View android.view.LayoutInflater.createViewFromTag(android.view.View, java.lang.String, android.content.Context, android.util.AttributeSet) (LayoutInflater.java:727)
        at void android.view.LayoutInflater.rInflate(org.xmlpull.v1.XmlPullParser, android.view.View, android.content.Context, android.util.AttributeSet, boolean) (LayoutInflater.java:858)
        at void android.view.LayoutInflater.rInflateChildren(org.xmlpull.v1.XmlPullParser, android.view.View, android.util.AttributeSet, boolean) (LayoutInflater.java:821)
        at android.view.View android.view.LayoutInflater.inflate(org.xmlpull.v1.XmlPullParser, android.view.ViewGroup, boolean) (LayoutInflater.java:518)
        at android.view.View android.view.LayoutInflater.inflate(int, android.view.ViewGroup, boolean) (LayoutInflater.java:426)
        at android.view.View android.view.LayoutInflater.inflate(int, android.view.ViewGroup) (LayoutInflater.java:377)
        at android.view.ViewGroup androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor() (AppCompatDelegateImpl.java:896)
        at void androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor() (AppCompatDelegateImpl.java:806)
        at void androidx.appcompat.app.AppCompatDelegateImpl.setContentView(int) (AppCompatDelegateImpl.java:693)
        at void androidx.appcompat.app.AppCompatActivity.setContentView(int) (AppCompatActivity.java:170)
        at void com.helio.app.MainActivity.onCreate(android.os.Bundle) (MainActivity.java:27)
        at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:6679)
        at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1118)
        at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2618)
        at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2726)
        at void android.app.ActivityThread.-wrap12(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
        at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1477)
        at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:102)
        at void android.os.Looper.loop() (Looper.java:154)
        at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6119)
        at java.lang.Object java.lang.reflect.Method.invoke!(java.lang.Object, java.lang.Object[]) (Method.java:-2)
        at void com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run() (ZygoteInit.java:886)
        at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:776)
    Caused by: java.lang.ClassNotFoundException: Didn't find class "android.view.View$OnUnhandledKeyEventListener" on path: DexPathList[[zip file "/data/app/com.helio.app-1/base.apk"],nativeLibraryDirectories=[/data/app/com.helio.app-1/lib/x86, /system/lib, /vendor/lib]]
        at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:56)
        at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:380)
        at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
        at void androidx.core.view.ViewCompat.setBackground(android.view.View, android.graphics.drawable.Drawable) (ViewCompat.java:2678)
        at void androidx.appcompat.widget.ActionBarContainer.<init>(android.content.Context, android.util.AttributeSet) (ActionBarContainer.java:63)
        at java.lang.Object java.lang.reflect.Constructor.newInstance0!(java.lang.Object[]) (Constructor.java:-2)
        at java.lang.Object java.lang.reflect.Constructor.newInstance(java.lang.Object[]) (Constructor.java:430)
        at android.view.View android.view.LayoutInflater.createView(java.lang.String, java.lang.String, android.util.AttributeSet) (LayoutInflater.java:645)
        at android.view.View android.view.LayoutInflater.createViewFromTag(android.view.View, java.lang.String, android.content.Context, android.util.AttributeSet, boolean) (LayoutInflater.java:787)
        at android.view.View android.view.LayoutInflater.createViewFromTag(android.view.View, java.lang.String, android.content.Context, android.util.AttributeSet) (LayoutInflater.java:727)
        at void android.view.LayoutInflater.rInflate(org.xmlpull.v1.XmlPullParser, android.view.View, android.content.Context, android.util.AttributeSet, boolean) (LayoutInflater.java:858)
        at void android.view.LayoutInflater.rInflateChildren(org.xmlpull.v1.XmlPullParser, android.view.View, android.util.AttributeSet, boolean) (LayoutInflater.java:821)
        at android.view.View android.view.LayoutInflater.inflate(org.xmlpull.v1.XmlPullParser, android.view.ViewGroup, boolean) (LayoutInflater.java:518)
        at android.view.View android.view.LayoutInflater.inflate(int, android.view.ViewGroup, boolean) (LayoutInflater.java:426)
        at android.view.View android.view.LayoutInflater.inflate(int, android.view.ViewGroup) (LayoutInflater.java:377)
        at android.view.ViewGroup androidx.appcompat.app.AppCompatDelegateImpl.createSubDecor() (AppCompatDelegateImpl.java:896)
        at void androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor() (AppCompatDelegateImpl.java:806)
        at void androidx.appcompat.app.AppCompatDelegateImpl.setContentView(int) (AppCompatDelegateImpl.java:693)
        at void androidx.appcompat.app.AppCompatActivity.setContentView(int) (AppCompatActivity.java:170)
        at void com.helio.app.MainActivity.onCreate(android.os.Bundle) (MainActivity.java:27)
        at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:6679)
        at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentation.java:1118)
        at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent) (ActivityThread.java:2618)
        at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:2726)
        at void android.app.ActivityThread.-wrap12(android.app.ActivityThread, android.app.ActivityThread$ActivityClientRecord, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
        at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1477)
        at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:102)
        at void android.os.Looper.loop() (Looper.java:154)
        at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:6119)
        at java.lang.Object java.lang.reflect.Method.invoke!(java.lang.Object, java.lang.Object[]) (Method.java:-2)
        at void com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run() (ZygoteInit.java:886)
        at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:776)
I/ViewConfigCompat: Could not find method getScaledScrollFactor() on ViewConfiguration
D/NetworkSecurityConfig: No Network Security Config specified, using platform default
I/okhttp.OkHttpClient: --> GET http://10.0.2.2:4310/motor/get_all
I/okhttp.OkHttpClient: --> END GET
I/TextToSpeech: Sucessfully bound to com.google.android.tts
I/okhttp.OkHttpClient: <-- 200 OK http://10.0.2.2:4310/motor/get_all (136ms)
    Content-Type: application/json
    Content-Length: 713
    Server: Werkzeug/1.0.1 Python/3.8.3
    Date: Tue, 16 Mar 2021 20:42:46 GMT
I/okhttp.OkHttpClient: [
      {
        "active": true,
        "battery": 0,
        "id": 0,
        "ip": "198.51.100.42",
        "length": 0,
        "level": 26,
        "name": "kitchen-1",
        "style": "res/drawable/ic_fridge.xml"
      },
I/okhttp.OkHttpClient:   {
        "active": true,
        "battery": 0,
        "id": 1,
        "ip": "198.51.100.42",
        "length": 0,
        "level": 59,
        "name": "kitchen-2",
        "style": "res/drawable/ic_fridge.xml"
      },
      {
        "active": true,
        "battery": 0,
        "id": 2,
        "ip": "198.51.100.42",
        "length": 0,
        "level": 84,
        "name": "bedroom",
        "style": "res/drawable/ic_bed.xml"
      },
      {
        "active": true,
        "battery": 0,
        "id": 3,
        "ip": "0.0.0.0",
        "length": 0,
        "level": 0,
        "name": "ddddd",
        "style": ""
      }
    ]
    <-- END HTTP (713-byte body)
D/: HostConnection::get() New Host Connection established 0xa03c8840, tid 4567
I/TextToSpeech: Connected to ComponentInfo{com.google.android.tts/com.google.android.tts.service.GoogleTTSService}
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@20ac041 succeeded: [Motor{id='0', name='kitchen-1', ip='198.51.100.42', active=true, level=26, battery=0, length=0, style='res/drawable/ic_fridge.xml', icon=null}, Motor{id='1', name='kitchen-2', ip='198.51.100.42', active=true, level=59, battery=0, length=0, style='res/drawable/ic_fridge.xml', icon=null}, Motor{id='2', name='bedroom', ip='198.51.100.42', active=true, level=84, battery=0, length=0, style='res/drawable/ic_bed.xml', icon=null}, Motor{id='3', name='ddddd', ip='0.0.0.0', active=true, level=0, battery=0, length=0, style='', icon=null}]
    Updated local state for [Motor{id='0', name='kitchen-1', ip='198.51.100.42', active=true, level=26, battery=0, length=0, style='res/drawable/ic_fridge.xml', icon=null}, Motor{id='1', name='kitchen-2', ip='198.51.100.42', active=true, level=59, battery=0, length=0, style='res/drawable/ic_fridge.xml', icon=null}, Motor{id='2', name='bedroom', ip='198.51.100.42', active=true, level=84, battery=0, length=0, style='res/drawable/ic_bed.xml', icon=null}, Motor{id='3', name='ddddd', ip='0.0.0.0', active=true, level=0, battery=0, length=0, style='', icon=null}]
I/TextToSpeech: Set up connection to ComponentInfo{com.google.android.tts/com.google.android.tts.service.GoogleTTSService}
D/: HostConnection::get() New Host Connection established 0x9189a040, tid 4605
I/OpenGLRenderer: Initialized EGL, version 1.4
D/OpenGLRenderer: Swap behavior 1
W/OpenGLRenderer: Failed to choose config with EGL_SWAP_BEHAVIOR_PRESERVED, retrying without...
D/OpenGLRenderer: Swap behavior 0
D/EGL_emulation: eglCreateContext: 0x9ff4e2c0: maj 2 min 0 rcv 2
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
I/art: Do partial code cache collection, code=18KB, data=30KB
    After code cache collection, code=16KB, data=29KB
    Increasing code cache capacity to 128KB
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
I/art: Do partial code cache collection, code=51KB, data=62KB
I/art: After code cache collection, code=51KB, data=62KB
    Increasing code cache capacity to 256KB
I/art: Do full code cache collection, code=117KB, data=118KB
I/art: Waiting for a blocking GC JitCodeCache
I/art: WaitForGcToComplete blocked for 8.622ms for cause JitCodeCache
    Starting a blocking GC JitCodeCache
    After code cache collection, code=86KB, data=64KB
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/start/2
    Content-Length: 0
I/okhttp.OkHttpClient: --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/update/2
    Content-Type: application/json; charset=UTF-8
    Content-Length: 121
I/okhttp.OkHttpClient: {"active":true,"battery":0,"ip":"198.51.100.42","length":0,"level":84,"name":"bedroom","style":"res/drawable/ic_bed.xml"}
    --> END PATCH (121-byte body)
I/okhttp.OkHttpClient: <-- 200 OK http://10.0.2.2:4310/motor/calibrate/start/2 (64ms)
I/okhttp.OkHttpClient: Content-Type: application/json
    Content-Length: 22
    Server: Werkzeug/1.0.1 Python/3.8.3
    Date: Tue, 16 Mar 2021 20:43:40 GMT
I/okhttp.OkHttpClient: "Calibration started"
I/okhttp.OkHttpClient: <-- END HTTP (22-byte body)
I/okhttp.OkHttpClient: <-- 200 OK http://10.0.2.2:4310/motor/update/2 (54ms)
    Content-Type: application/json
    Content-Length: 162
I/okhttp.OkHttpClient: Server: Werkzeug/1.0.1 Python/3.8.3
    Date: Tue, 16 Mar 2021 20:43:40 GMT
    {
      "active": true,
      "battery": 0,
      "id": 2,
      "ip": "198.51.100.42",
      "length": 0,
      "level": 84,
      "name": "bedroom",
      "style": "res/drawable/ic_bed.xml"
    }
    <-- END HTTP (162-byte body)
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@ec1353b succeeded: Response{protocol=http/1.0, code=200, message=OK, url=http://10.0.2.2:4310/motor/calibrate/start/2}
    retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@d1dd458 succeeded: Motor{id='2', name='bedroom', ip='198.51.100.42', active=true, level=84, battery=0, length=0, style='res/drawable/ic_bed.xml', icon=null}
    Updating local state for Motor{id='2', name='bedroom', ip='198.51.100.42', active=true, level=84, battery=0, length=0, style='res/drawable/ic_bed.xml', icon=null}
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/move_up/2
    Content-Length: 0
    --> END PATCH (0-byte body)
D/EGL_emulation: eglMakeCurrent: 0x9ff4e2c0: ver 2 0 (tinfo 0xa21a2e80)
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/stop_moving/2
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- HTTP FAILED: java.net.SocketTimeoutException: timeout
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@b696a6e failed: java.net.SocketTimeoutException: timeout
I/okhttp.OkHttpClient: <-- HTTP FAILED: java.net.SocketTimeoutException: timeout
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@964050f failed: java.net.SocketTimeoutException: timeout
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/set_highest/2
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/move_down/2
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/stop_moving/2
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- HTTP FAILED: java.net.SocketTimeoutException: timeout
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@4e5e32b failed: java.net.SocketTimeoutException: timeout
I/okhttp.OkHttpClient: <-- HTTP FAILED: java.net.SocketTimeoutException: timeout
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@f5b0388 failed: java.net.SocketTimeoutException: timeout
I/okhttp.OkHttpClient: <-- HTTP FAILED: java.net.SocketTimeoutException: timeout
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@8042121 failed: java.net.SocketTimeoutException: timeout
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/set_lowest/2
I/okhttp.OkHttpClient: Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- HTTP FAILED: java.net.SocketTimeoutException: timeout
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@b8c3d2 failed: java.net.SocketTimeoutException: timeout
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/stop_moving/2
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: --> PATCH http://10.0.2.2:4310/motor/calibrate/stop/2
    Content-Length: 0
    --> END PATCH (0-byte body)
I/okhttp.OkHttpClient: <-- 200 OK http://10.0.2.2:4310/motor/calibrate/stop/2 (13ms)
    Content-Type: application/json
    Content-Length: 23
    Server: Werkzeug/1.0.1 Python/3.8.3
    Date: Tue, 16 Mar 2021 20:45:18 GMT
I/okhttp.OkHttpClient: "Calibration finished"
    <-- END HTTP (23-byte body)
I/art: Do partial code cache collection, code=112KB, data=112KB
    After code cache collection, code=112KB, data=112KB
    Increasing code cache capacity to 512KB
I/System.out: retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall@a3d8043 succeeded: Response{protocol=http/1.0, code=200, message=OK, url=http://10.0.2.2:4310/motor/calibrate/stop/2}
```

And on the hub side:
```
python api.py 
*********************  THREADS  **********************
MainThread
SchedulerThread
******************************************************
 * Serving Flask app "api" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: on
 * Running on http://0.0.0.0:4310/ (Press CTRL+C to quit)
 * Restarting with stat
*********************  THREADS  **********************
MainThread
SchedulerThread
******************************************************
 * Debugger is active!
 * Debugger PIN: 245-606-390
127.0.0.1 - - [16/Mar/2021 21:42:46] "GET /motor/get_all HTTP/1.1" 200 -
127.0.0.1 - - [16/Mar/2021 21:43:40] "PATCH /motor/calibrate/start/2 HTTP/1.1" 200 -
127.0.0.1 - - [16/Mar/2021 21:43:40] "PATCH /motor/update/2 HTTP/1.1" 200 -
127.0.0.1 - - [16/Mar/2021 21:45:18] "PATCH /motor/calibrate/stop/2 HTTP/1.1" 200 -
HTTPConnectionPool(host='198.51.100.42', port=4310): Max retries exceeded with url: /calibration_move_up (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8da2a295e0>: Failed to establish a new connection: [Errno 110] Connection timed out'))
127.0.0.1 - - [16/Mar/2021 21:46:21] "PATCH /motor/calibrate/move_up/2 HTTP/1.1" 500 -
HTTPConnectionPool(host='198.51.100.42', port=4310): Max retries exceeded with url: /calibration_stop_moving (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8da2a41be0>: Failed to establish a new connection: [Errno 110] Connection timed out'))
127.0.0.1 - - [16/Mar/2021 21:46:25] "PATCH /motor/calibrate/stop_moving/2 HTTP/1.1" 500 -
HTTPConnectionPool(host='198.51.100.42', port=4310): Max retries exceeded with url: /calibration_set_highest (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8da19c6f10>: Failed to establish a new connection: [Errno 110] Connection timed out'))
127.0.0.1 - - [16/Mar/2021 21:46:39] "PATCH /motor/calibrate/set_highest/2 HTTP/1.1" 500 -
HTTPConnectionPool(host='198.51.100.42', port=4310): Max retries exceeded with url: /calibration_move_down (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8da19d4340>: Failed to establish a new connection: [Errno 110] Connection timed out'))
127.0.0.1 - - [16/Mar/2021 21:46:47] "PATCH /motor/calibrate/move_down/2 HTTP/1.1" 500 -
HTTPConnectionPool(host='198.51.100.42', port=4310): Max retries exceeded with url: /calibration_stop_moving (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8da19dd550>: Failed to establish a new connection: [Errno 110] Connection timed out'))
127.0.0.1 - - [16/Mar/2021 21:46:49] "PATCH /motor/calibrate/stop_moving/2 HTTP/1.1" 500 -
HTTPConnectionPool(host='198.51.100.42', port=4310): Max retries exceeded with url: /calibration_set_lowest (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8da19e3910>: Failed to establish a new connection: [Errno 110] Connection timed out'))
127.0.0.1 - - [16/Mar/2021 21:47:14] "PATCH /motor/calibrate/set_lowest/2 HTTP/1.1" 500 -
HTTPConnectionPool(host='198.51.100.42', port=4310): Max retries exceeded with url: /calibration_stop_moving (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8da19f1790>: Failed to establish a new connection: [Errno 110] Connection timed out'))
127.0.0.1 - - [16/Mar/2021 21:47:28] "PATCH /motor/calibrate/stop_moving/2 HTTP/1.1" 500 -
```

NB: there is a 1min pause following `127.0.0.1 - - [16/Mar/2021 21:45:18] "PATCH /motor/calibrate/stop/2 HTTP/1.1" 200 -`

adamb56789 commented:
Removing the sleep had no effect. I think this is probably a hub side issue, and something do to with the fact that it can't communicate with the motor server. My guess is that we must have different network configurations or something which is somehow causing the hub to behave in a different manner - I am getting `[WinError 10049] The requested address is not valid in its context` and you are getting `[Errno 110] Connection timed out`. Actually it is probably because different operating systems. Also note that in both of our hub errors the url listed (e.g. `/calibration_move_up`) is the one that it using to try to connect to the motor server.

nout-kleef commented on app/src/main/java/com/helio/app/UserDataViewModel.java:
```diff
@@ -109,24 +108,8 @@ public void setHubIp(String ip) {
-        );
-        client.updateMotor(motors, currentMotorId, motorSettingsRequest);
+        // This kind of operation is deprecated and no longer available
+        throw new UnsupportedOperationException();
```
Can we just remove this method and the "open" and "close" method instead? That seems cleaner to me, because this causes the app to crash when either of those buttons is pressed

adamb56789 replied:
This is removed in the blinds settings redesign, so that was just a temporary measure.
